### PR TITLE
[Deepseek Blitz] Split q a proj mm on inner dim

### PR DESCRIPTION
**Pull request by aagarwalTT**
### Ticket
Closes: https://github.com/tenstorrent/tt-metal/issues/36975

### Problem description
Provide context for the problem.

### What's changed
Sharded the MM weights into two core groups, such that each 48 core group receives half of the original MM weights. Each core based on its logical coords then decides which half of the MM inputs to operate on. 

Added 1 new CB for now - for gather output (half1)

Added a new gather reduce op that does a reduction step after the gather senders are done finish sending. 

Update MM1 to be KNSlicedMatmul instead of regular matmul

### Checklist

- [ ] [All post-commit tests](https://github.com/tenstorrent/tt-metal/actions/workflows/all-post-commit-workflows.yaml?query=branch:aagarwal/split-q-a-proj-mm)
- [ ] [Blackhole Post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-post-commit.yaml?query=branch:aagarwal/split-q-a-proj-mm)
- [ ] [cpp-unit-tests](https://github.com/tenstorrent/tt-metal/actions/workflows/tt-metal-l2-nightly.yaml?query=branch:aagarwal/split-q-a-proj-mm)
- [ ] New/Existing tests provide coverage for changes

#### Model tests

If your changes cover model-related code, you should run tests corresponding to affected models and platforms (Single card, T3K, Galaxy). "Choose your pipeline" workflows facilitate running multiple kinds of tests in a single run. Each offers `models-mandatory` and `models-extended` presets.
The former includes a minimal set of tests, to be run always. The latter extends that with additional ones - use your best judgement in deciding which is the most appropriate for your PR.

- [ ] [(Single) Choose your pipeline](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select.yaml?query=branch:aagarwal/split-q-a-proj-mm)
  - [ ] `models-mandatory` preset (runs: [Device perf regressions](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-device-models.yaml) and [Frequent model and ttnn tests](https://github.com/tenstorrent/tt-metal/actions/workflows/fast-dispatch-full-regressions-and-models.yaml))
  - [ ] `models-extended` preset (runs: the mandatory tests, plus [Demo](https://github.com/tenstorrent/tt-metal/actions/workflows/single-card-demo-tests.yaml) and [Model perf](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-models.yaml) tests)
  - [ ] other selection - specify runs

- [ ] [(T3K) Choose your pipeline](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-t3k.yaml?query=branch:aagarwal/split-q-a-proj-mm)
  - [ ] `models-mandatory` preset (runs: [Unit tests](https://github.com/tenstorrent/tt-metal/actions/workflows/t3000-unit-tests.yaml))
  - [ ] `models-extended` preset (runs: the mandatory tests, plus [Demo](https://github.com/tenstorrent/tt-metal/actions/workflows/t3000-demo-tests.yaml) and [Model perf](https://github.com/tenstorrent/tt-metal/actions/workflows/t3000-model-perf-tests.yaml) tests)
  - [ ] other selection - specify runs

- [ ] [(Galaxy) Choose your pipeline](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-galaxy.yaml?query=branch:aagarwal/split-q-a-proj-mm)
  - [ ] `models-mandatory` preset (runs: [Quick tests](https://github.com/tenstorrent/tt-metal/actions/workflows/galaxy-quick.yaml))
  - [ ] `models-extended` preset (runs: the mandatory tests, plus [Demo](https://github.com/tenstorrent/tt-metal/actions/workflows/galaxy-demo-tests.yaml) and [Model perf](https://github.com/tenstorrent/tt-metal/actions/workflows/galaxy-perf-tests.yaml) tests)
  - [ ] other selection - specify runs

### CODE REVIEW

File: models/demos/deepseek_v3_b1/unified_kernels/gather_reduce.hpp [hunk 1]
```diff
@@ -0,0 +1,215 @@
...
+            // ================================================================
+            // NCRISC (Sender) - DataMovementProcessor.RISCV_1
+            // ================================================================
+            if constexpr (IsSenderCore) {
+                bool is_half0 = (my_logical_x_ < args.matmul_half_boundary_col);
+                uint32_t dst_cb_id = is_half0 ? args.half0_cb_id : args.half1_cb_id;
+                uint32_t dst_base_addr = get_write_ptr(dst_cb_id);
+
+                uint32_t local_x = is_half0 ? my_logical_x_ : (my_logical_x_ - args.matmul_half_boundary_col);
+                uint32_t sender_idx = local_x + my_logical_y_ * args.matmul_cols_per_half;
```
[the remark this answers is not in the document]
Comment (aagarwalTT) [reply models/demos/deepseek_v3_b1/unified_kernels/gather_reduce.hpp]: fixed

File: models/demos/deepseek_v3_b1/fused_ops/pre_sdpa/op.py [hunk 2]
```diff
@@ -295,17 +298,46 @@ def op(
...
+
+        matmul_weights_core_grid = matmul_half0_core_grid.merge(matmul_half1_core_grid)
+        matmul_bbox = matmul_weights_core_grid.bounding_box()
+        matmul_grid_size = matmul_bbox.grid_size()
+        if matmul_grid_size.x % 2 != 0:
+            raise ValueError(f"matmul core grid x-dimension must be even for half split, got {matmul_grid_size.x}")
+        if matmul_grid_size.x * matmul_grid_size.y != 96:
+            raise ValueError(
+                f"matmul core grid must have 96 cores for this K-split path, got {matmul_grid_size.x * matmul_grid_size.y}"
+            )
```
[the remark this answers is not in the document]
Comment (aagarwalTT) [reply models/demos/deepseek_v3_b1/fused_ops/pre_sdpa/op.py]: fixed

File: models/demos/deepseek_v3_b1/fused_ops/pre_sdpa/kernels/pre_sdpa_kernel.cpp [hunk 3]
```diff
@@ -101,28 +103,32 @@ void kernel_main() {
...
 
     // Matmul CTArgs type alias (NCRISC uses ReaderCTArgs)
-    using MatmulCTArgs = deepseek_b1_ops::Matmul::ReaderCTArgs;
+    using MatmulCTArgs = deepseek_b1_ops::KNSlicedMatmul::ReaderCTArgs;
     using Matmul2CTArgs = deepseek_b1_ops::Matmul::ReaderCTArgs;
     using Matmul3CTArgs = deepseek_b1_ops::Matmul::ReaderCTArgs;
 
     // Matmul reader args (NCRISC is no-op)
-    deepseek_b1_ops::Matmul::ReaderArgs matmul_args{};
+    deepseek_b1_ops::KNSlicedMatmul::ReaderArgs matmul_args{};
```
Comment (TT-BrianLiu) [inline models/demos/deepseek_v3_b1/fused_ops/pre_sdpa/kernels/pre_sdpa_kernel.cpp:111]: What is KNSlicedMatmul? Are you missing a file somewhere in the commit?

Comment (TT-BrianLiu) [reply models/demos/deepseek_v3_b1/fused_ops/pre_sdpa/kernels/pre_sdpa_kernel.cpp:111]: nvm it's already on main

File: models/demos/deepseek_v3_b1/tests/unit_tests/test_pre_sdpa.py [hunk 4]
```diff
@@ -279,23 +281,53 @@ def test_pre_sdpa(
...
+    num_matmul_half_cores = (matmul_grid_x // 2) * matmul_grid_y
+    # WIDTH_SHARDED requires shard height == tensor height.
+    # Each half tensor is [3584, 1536], so per-core shard is [3584, 32].
+    matmul_half_shard_shape = (
+        matmul_weights_shape[0] // 2,
+        matmul_weights_shape[1] // num_matmul_half_cores,
+    )
+    matmul_half0_shard_spec = ttnn.ShardSpec(
+        ttnn.CoreRangeSet({matmul_half0_grid}),
+        matmul_half_shard_shape,
```
Comment (TT-BrianLiu) [inline models/demos/deepseek_v3_b1/tests/unit_tests/test_pre_sdpa.py]: Can we just reshuffle the weights into [3584, 3072] and just do width sharding over the 96 cores? Keep the single cb throughout

**Review by TT-BrianLiu — APPROVED** [review]
Looks good. One comment about how to shard the input weights. I think keeping it as one is cleaner and easier to stitch

File: models/demos/deepseek_v3_b1/tests/unit_tests/test_pre_sdpa.py (hunk 4 above)
Comment (TT-BrianLiu) [reply models/demos/deepseek_v3_b1/tests/unit_tests/test_pre_sdpa.py]: the dims come from [7168 / 2, 1536 * 2]

File: models/demos/deepseek_v3_b1/unified_kernels/gather_reduce.hpp [hunk 5]
```diff
@@ -0,0 +1,220 @@
...
+        void operator()(const RTArgs& args) { impl(args); }
+
+    private:
+        void impl([[maybe_unused]] const RTArgs& args) {
+#if defined(COMPILE_FOR_NCRISC)
+            // ================================================================
+            // NCRISC (Sender) - DataMovementProcessor.RISCV_1
+            // ================================================================
+            if constexpr (IsSenderCore) {
+                const auto half_info = unified_kernels::get_split_half_core_info<true>(
```
Comment (TT-BrianLiu) [inline models/demos/deepseek_v3_b1/unified_kernels/gather_reduce.hpp:151]: I might have overcomplicated things when I updated the original gather to just query offset from the grid. I believe it's modified to just pass in the offset addr (which is what it was before). But here, this is ok to keep since it's still some nice grid. But just FYI that this is not something we necessarily need if it gets annoying to maintain

File: models/demos/deepseek_v3_b1/tests/unit_tests/test_pre_sdpa.py [hunk 6]
```diff
@@ -179,6 +179,9 @@ def test_pre_sdpa(
     torch_input = torch.randn(shape, dtype=torch.bfloat16)
     torch_gamma = torch.randn(shape, dtype=torch.bfloat16)
     torch_matmul_weights = torch.randn(matmul_weights_shape, dtype=torch.bfloat16)
+    torch_matmul_weights_half0 = torch_matmul_weights[: matmul_weights_shape[0] // 2, :]
+    torch_matmul_weights_half1 = torch_matmul_weights[matmul_weights_shape[0] // 2 :, :]
+    torch_matmul_weights_packed = torch.cat([torch_matmul_weights_half0, torch_matmul_weights_half1], dim=1)
```
Comment (TT-BrianLiu) [inline models/demos/deepseek_v3_b1/tests/unit_tests/test_pre_sdpa.py]: Maybe something like this would be cleaner? No real preference either way
```
torch_matmul_weights_packed = (
    torch_matmul_weights
        .reshape(2, H // 2, W)
        .permute(1, 0, 2)
        .reshape(H // 2, 2 * W)
)
```

Comment (aagarwalTT) [reply models/demos/deepseek_v3_b1/tests/unit_tests/test_pre_sdpa.py]: done
